proposed/2021/PackageNamespaces.md

```diff
@@ -0,0 +1,568 @@
+
+**Definition:**
+
+Add a new `<packageNamespaces>` element within the `NuGet.config` using the following syntax:
+
+```xml
+<packageSources>
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
+         
+    <add key="contoso" value="https://contoso.com/packages/" />
+ 
+</packageSources>
+ 
+<packageNamespaces>
+</packageNamespaces>
+```
+
+Define the `<packageSource>` element within the `<packageNamespaces>` parent with the name of a valid package source:
+
+```xml
+<packageSources>
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
+         
+    <add key="contoso" value="https://contoso.com/packages/" />
+ 
+</packageSources>
+ 
+<packageNamespaces>
+    <packageSource name="nuget.org">
+
+    </packageSource>
+ 
+</packageNamespaces>
+```
+
+Next, add `<namespace>` elements under the `<packageSource>` with an `id` to specify the namespace. Namespaces can be defined as their full namepsace or using a wildcard(*) to match the glob pattern of 0 or more package ID(s):
+
+```xml
+<packageSources>
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
+         
+    <add key="contoso" value="https://contoso.com/packages/" />
+ 
+</packageSources>
+ 
+<packageNamespaces>
+    <!-- Add a namespace for Microsoft.* or NuGet.* under the nuget.org source. -->
+    <packageSource name="nuget.org">
+        <namespace id="Microsoft.*" />
+        <namespace id="NuGet.*" />
+    </packageSource>
+ 
+</packageNamespaces>
+```
+
+If a user would like to support namespaces on different package sources, they can repeat this process for each source.
+A user may also pin a specific package id instead of the complete namespace.
+Specific ids take precedence over the namespaces.
+
+**Example:**
+
+```xml
+<packageSources>
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" protocolVersion="3" />
+         
+    <add key="contoso" value="https://contoso.com/packages/" />
+ 
+</packageSources>
+ 
+<packageNamespaces>
+    <!-- Add a namespace for Microsoft.* or NuGet.* under the nuget.org source. -->
+    <packageSource name="nuget.org">
+        <namespace id="Microsoft.*" />
+        <namespace id="NuGet.*" />
+    </packageSource>
+
+    <!-- Add a namespace for Contoso.* under the contoso source. -->
+    <packageSource name="contoso">
+        <namespace id="Contoso.*" />
+        <!-- Add a specific id to be download from  the contoso source. -->
+        <namespace id="Special.Package.With.An.Imperfect.Id" />
+    </packageSource>
+ 
+</packageNamespaces>
+```
+
+By default, namespaces will allow a package to match multiple namespaces and download with no precedence. In some cases, one might want to avoid ambiguous cases. That can be done by adding a `strict` flag.
+
+```xml
+<config>
+    <add key="namespaceMode" value="strict" />
```

Comment:
I'm not a fan of "strict" here. It doesn't really tell me much about what that means, and it's already a bit of a confusing feature to think about.

Reply:
Do you have a recommendation here? 

signatureValidationMode  has `require` and `allow` which I don't love either. 
`strict` matches a feature that the compiler has.

Reply:
Maybe rename the option away from `signatureValidationMode` (which doesn't tell us much), to something more descriptive like `multipleNameSpaceSources`, and then it can just be `true` and `false`?

Reply:
I also think it's a bit confusing, and perhaps not all that useful as it's spec'd out here? The issue is that I think it would be hard to turn on in any case where multiple feeds are present. Consider .NET. We build plenty of pre-release packages internally, and also put them on nuget.org. Thus I would want to put Microsoft.* namespaces on both my nuget.org and dotnet6 feed, and that would be completely valid. It sounds like strict mode would disable that ability?

Perhaps instead, the strict mode would specify that all package id's must be accounted for? E.g. If I have 3 feeds:
- nuget.org
- dotnet-eng
- dotnet6

Let's say I said that `NewtonSoft.*` comes from nuget.org and `Microsoft.` comes from dotnet-eng or dotnet6. If I then try to reference `Xunit.Whatever`, the restore would fail in strict mode because that package ID/namespace was not assigned to any feed.

Reply:
Should strict mode imply both, so either a 1-1 relationship between package id and source, or a 1-n *explicit* relationship between package id & source.

I understand the special scenario for the .NET repos, and that's certainly a big consideration, but we also expect the more common scenario to be 1 private feed & multiple additional feeds.

Reply:
> Should strict mode imply both, so either a 1-1 relationship between package id and source, or a 1-n explicit relationship between package id & source.

How would that be implemented? I'm not sure I follow.

> I understand the special scenario for the .NET repos, and that's certainly a big consideration, but we also expect the more common scenario to be 1 private feed & multiple additional feeds.

I don't think .NET's scenario is all that special, but can we get some sense of that from the community? I think you're likely to see ID overlap in any case where pre-release packages are placed on a feed and verified, prior to publishing to nuget.org.

Reply:
> How would that be implemented? I'm not sure I follow.

Maybe some examples clarify it: 

The packages are: 

Microsoft.* and NuGet.*. 

* Case 1*

Source 1: Microsoft.*
Source 2: Microsoft.*, NuGet.*

This can be a valid strict mode scenario. 

* Case 2:*

Source 1: Microsoft.*
Source 2: * 

This a valid strict mode scenario. 

*Case 3:*
Source 1: NuGet.*
Source 2: *
Source 3: *

This is invalid because you are not explicitly opting into using both sources for the specific packages.